### PR TITLE
Remove logic to skip checkpoint save if checkpoint exists

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by mikolajblaz — DISMISSED** [review]

**Comment by maanug-nv** [comment]
based on current test failure, looks like this requires changes in MCore as well to handle and existing dist ckpt dir. i'd say we should make that change first before merging this

**Comment by mikolajblaz** [comment]
> based on current test failure, looks like this requires changes in MCore as well to handle and existing dist ckpt dir. i'd say we should make that change first before merging this

How can I check the failing tests? The only failing job I see is this one but it doesn't have any useful info: https://github.com/NVIDIA/NeMo/actions/runs/11958751953/job/33339441430?pr=11362

Regardless of the failure, I would recommend ensuring that on the NeMo side the checkpointing logic ensures that we don't save twice to the same directory, otherwise that could lead to silent errors.

We did it for NeMo 1.0 and currently all such circumstances (saving twice to the same dir) are explicitly treated as an error which I believe leads to a better experiments handling from a user perspective

**Comment by maanug-nv** [comment]
from that link, go to the Summary, and see the 'Failed jobs'. 
> I would recommend ensuring that on the NeMo side the checkpointing logic ensures that we don't save twice to the same directory, otherwise that could lead to silent errors.

You're right, I completely agree. Currently on main, async save will crash if the dir exists. Non-async save will just print warning and skip saving. Looks like this change is exposing that problem in one of the tests. 

What I don't understand is why we aren't running the PTL logic that appends a version counter ([1](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/callbacks/model_checkpoint.py#L661-L665), [2](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/callbacks/model_checkpoint.py#L685-L689))

**Comment by ashors1** [comment]
>Currently on main, async save will crash if the dir exists. Non-async save will just print warning and skip saving. Looks like this change is exposing that problem in one of the tests.

With this PR, both async and synchronous checkpointing will error if we try to save to an existing directory, which I think is what we want, right? In that case, maybe we can just update the test to save to different directories

**Comment by maanug-nv** [comment]
ran the failing test locally with your changes, I think FSDPStrategy is missing `import os` so removing previous last checkpoint will fail.

**Comment by ashors1** [comment]
> ran the failing test locally with your changes, I think FSDPStrategy is missing `import os` so removing previous last checkpoint will fail: https://github.com/NVIDIA/NeMo/blob/main/nemo/lightning/pytorch/strategies/fsdp_strategy.py#L198-L205

Hm yeah, I noticed that and tried to update but looks like there was a problem with pushing my changes. Let me try again

**Review by maanug-nv — APPROVED** [review]

**Review by maanug-nv — APPROVED** [review]

File: nemo/lightning/io/pl.py [hunk 1]
```diff
@@ -155,9 +155,6 @@ def save_checkpoint(self, checkpoint: Dict[str, Any], path: _PATH, storage_optio
         checkpoint_dir = ckpt_to_weights_subdir(path, is_saving=True)
 
         fs = get_filesystem(checkpoint_dir)
-        if fs.isdir(checkpoint_dir) and dist_checkpointing.check_is_distributed_checkpoint(checkpoint_dir):
-            logging.info(f'Distributed checkpoint at path {checkpoint_dir} already exists, skipping saving')
-            return
```
Comment (ananthsub) [inline nemo/lightning/io/pl.py:160]: @ashors1 @maanug-nv n00b question: could you describe why this is a fix? is it unexpected behavior that we want to raise more visibly? or something else?

Comment (maanug-nv) [reply nemo/lightning/io/pl.py:160]: > Currently on main, async save will crash if the dir exists. Non-async save will just print warning and skip saving. Looks like this change is exposing that problem in one of the tests.
> 
> With this PR, both async and synchronous checkpointing will error if we try to save to an existing directory, which I think is what we want, right? In that case, maybe we can just update the test to save to different directories

This change makes the error visible for synchronous and makes the error message more clear for both. The MCore dist_checkpointing library will raise the error now for both of these cases.

Comment (ashors1) [reply nemo/lightning/io/pl.py:160]: Yes, @mikolajblaz removed this in NeMo 1 and asked us to do the same in NeMo 2. We should not attempt to save two checkpoints to the same directory, and if a user ever tries, we'd like to throw an error rather than just skipping saving.
